### DISCUSSION TITLE
Using both cores of ESP32 DevkitC V4

### DISCUSSION THREAD

**Discussion by PWSJA** [micropython/micropython#10604] [ESP32] [answered]
MicroPython version: v1.19.1, downloaded from [here](https://micropython.org/download/esp32/)
Board:  ESP32 DevkitC V4
I use Visual Studio Code with the official Pymakr extension

Hello everyone,

In my project, I use a 4 bit 7 segmentdisplay (without a driver like TM1637). Therefore, my code has to run continuously to show something on the segmentdisplay. This code works fine, and the segmentdisplay shows its data with no visible flickering.

The ESP32 makes a webrequest every hour. To make the segment display continuously show its data, I use the _thread module. I watched this [video](https://www.youtube.com/watch?v=QeDnjcdGrpY), which told me that the _thread module allows me to use both cores of my ESP32. The problem is, that when I start one thread for the segmentdisplay and one to make a webrequest, the segmentdisplay starts to flicker, which means that the code that controls it is interrupted. 

I think that means both thread are running on the same core and the ESP32 switches between the threads, resulting in a flickering of the segmentdisplay. Is it possible to use both cores on the ESP32 and how?

```py
import _thread


import network_connection
from segment_display import SegmentDisplay


segment = SegmentDisplay(digits=(16, 18, 19, 4), segments=(5, 21, 26, 25, 33, 17, 27), show=False, wait_time=0.001) # intitializes the segment display
_thread.start_new_thread(segment.show, ())
data = network_connection.web_request() #  establishes a wificonnection and does a webrequest
```

Any help would be appreciated.

**Comment by karfas** [comment]
I am pretty sure this is caused the global interpreter lock (GIL) used in most python/micropython implementations.
For myself, I would simply tolerate a little flickering once an hour or so.
It **might** be possible that you get better results using asyncio (at least in the network thread). Asyncio does a quite good job to avoid lengthy/blocking system calls.

**Reply by PWSJA** [reply]
Firstly thank you for your reply!

So you're saying that is is not possible to use both the cores of the ESP32 with micropython? On the raspberry pi pico it seems to be possible, according to various people (among others [this](https://microcontrollerslab.com/dual-core-raspberry-pi-pico-programming-micropython/) one.) 

I first tried it with async but that didn't work really good because the webrequest held up te code of the segmentdisplay.

**Reply by karfas** [reply]
Please **read** about th GIL.
AFAIK, the RP2040 port is an exception (threads without GIL).

For your problem, you should find out what exactly leads to "flicker". 
I can't imagine that interruptions of the display code in the range of a few milliseconds are really visible.

**Reply by PWSJA** [reply]
Thank you for your reply. I will have a look at the code to find out what causes the flicker.

**Comment by jimmo** [comment, marked as answer]
> Is it possible to use both cores on the ESP32 and how?

@PWSJA MicroPython does not support dual-core on the ESP32, it will always run all its threads on the same core.

@karfas This isn't specifically related to the GIL -- the threads are on the same core the OP's problem would happen regardless.

You're right though, if we did add support for multi-core on ESP32 then we would have to disable the GIL in order to get any real benefit -- exactly like on rp2040.

**Comment by skrat** [comment]
I'm just leaving this here: to utilize the 2nd core (core 0, micropython runs on core 1 on dual core esp32), you can create user C module, use FreeRTOS API to create a task and pin it to core 0. Recommended reading: https://docs.espressif.com/projects/esp-idf/en/stable/esp32s3/api-reference/system/freertos_idf.html

I haven't tried this yet, but you could share a buffer between micropython and your task, ie. export it in your C module, and read/write to it from within the task that you pinned to core0.

**Comment by sosi-deadeye** [comment]
Is it possible to expose basic functionality of `freeRTOS` via Micropython for threads?

**Reply by skrat** [reply]
I don't think that would be practical in any way. FreeRTOS tasks should really only be C/C++/asm code. You could run another micropython instance on core0 and somehow coordinate the two, but that would be ton of work with diminishing returns. Even creating producer/consumer pattern where you push tasks to core0 would be challenging. Per ESP-IDF documentation, the idea is to run system tasks on core0 (wifi, ble, etc.), and user/app tasks (GUI) on core1.
